src/heartbeat.js

```diff
@@ -0,0 +1,185 @@
+import EventEmitter from 'events';
+import process from 'process';
+
+const DEFAULT_SETTINGS = {
+  workerInterval: 500,
+  monitorInterval: 1000,
+  stallTolerance: 5000,
+};
+
+function hrtimeToMilliseconds(hrtime) {
+  return Math.floor((hrtime[0] * 1e3) + (hrtime[1] / 1e6));
+}
+
+class HeartbeatMonitor extends EventEmitter {
+  constructor(supervisor, { monitorInterval, stallTolerance }) {
+    super();
+
+    this.supervisor = supervisor;
+    this.monitorInterval = monitorInterval || DEFAULT_SETTINGS.monitorInterval;
+    this.stallTolerance = stallTolerance || DEFAULT_SETTINGS.stallTolerance;
+
+    this.monitorHeartbeatTimeout = null;
+    this.heartbeatStartedAt = null;
+    this.timestamps = {};
+
+    this.recordHeartbeat = this.recordHeartbeat.bind(this);
```

Comment:
all the binds are only needed if you're passing them directly and you want them to stay bound - that doesn't seem to be the case for any of these.

Reply:
Is that not the case for callbacks passed to `setTimeout` e.g. https://github.com/airbnb/pool-hall/pull/1/files#diff-03be54f511190ff10f1e375c23bc60c6R73 ?

It looks like `recordHeartbeat` is never used as a callback by itself, but `start` and `monitorHeartbeat` are

Reply:
I verified that tests fail due to wandering `this` when I comment out the binds for `monitorHeartbeat` and `start`.  While testing this, I also added a case to specifically test catching stalls from monitoring called via the timer, rather than just the initial check.

Reply:
ah yes, i did miss that.

Either way tho, it seems like the binding should be done at the location of the setTimeout - external code should never assume that a method is autobound.